plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -314,18 +317,70 @@ checkTypeFromBinding recurs = \case
                       -- OPTIMIZE: we use inferKind for scope-checking, but a simple ADT-traversal would suffice
                       for_ (funTyArgs ty) inferKindM
 
+-- | Check that the in-Term's inferred type of a Let has kind *.
+-- Skip this check at the top-level, to allow top-level types to escape; see Note [PIR vs paper FIR Difference-2].
+checkStarInferred :: AsTypeError e term uni ann
+                  => ann -> Normalized (Type TyName uni b) -> TypeCheckM uni IsTop e ()
+checkStarInferred ann t = do
+    isTop <- asks view tceOther
+    case isTop of
+        NoTop  -> checkKindM ann (ann <$ unNormalized t) $ Type ()
+        -- NOTE: we completely skip the check in case of toplevel because we would need an *final, extended Gamma environment*
+        -- to run the kind-check in, but we cannot easily get that since we are using a Reader for environments and not State
+        YesTop -> pure ()
+
+-- | Extra flag to be passed in the TypeCheckM Reader context,
+-- to signal if the PIR expression currently being typechecked is at the top-level
+-- and thus its type can escape, or nested and thus not allowed to escape.
+data IsTop = YesTop | NoTop
+
+-- | Changes the flag in nested-lets so to disallow returning a type outside of the type's scope
+noEscapeType :: TypeCheckM uni IsTop e a -> TypeCheckM uni IsTop e a
+noEscapeType = local $ tceOther .~ NoTop
+
+-- | Run a 'TypeCheckM' computation by supplying a 'TypeCheckConfig' to it.
+-- Differs from its PLC version in that is passes an extra env flag 'YesTop'.
+runTypeCheckM :: (MonadError e m, MonadQuote m)
+              => TypeCheckConfig uni -> TypeCheckM uni IsTop e a -> m a
+runTypeCheckM config a =
+    liftEither =<< liftQuote (runExceptT $ runReaderT a env) where
+        env = TypeCheckEnv config mempty mempty YesTop
+
 -- Helpers
 ----------
 
+-- | Infer the type of a term.
+inferType
```

Comment:
Given what I said earlier (I think it's fine to expose the functions with the full config argument), I don't think we need to duplicate specialized versions of these, it's fine to use the generic ones from PLC.